data/transactions/logic/assembler_test.go

```diff
@@ -2938,6 +2942,330 @@ int 1
+#define abi-return  ;pushbytes 0x151f7c75; swap; concat; log; int 1; return;
+
+// stanza: "set $var from-{type}"
+#define parse ; int
+#define read_arg ;dup; txnas ApplicationArgs;
+#define from-string	;read_arg; abi-decode-bytes;  stores;
+#define from-uint16	;read_arg; abi-decode-uint16;  stores;
+#define from-uint32 ;read_arg; abi-decode-uint32;  stores;
+
+// stanza: "reply $var as-{type}
+#define returns ; int
+#define as-uint32; loads; abi-encode-uint32; abi-return;
+#define as-string; loads; abi-encode-bytes; abi-return;
+
+// Contract
+
+// echo handler
+method "echo(string)string"; abi-route echo
+echo:
+	#define msg 1
+	parse msg from-string
+
+	// cool things happen ...
+
+	returns msg as-string
+
+
+// add handler
+method "add(uint32,uint32)uint32"; abi-route add 
+add:
+	#define x 1
+	parse x from-uint32 
+
+	#define y 2
+	parse y from-uint32
+
+	#define sum 255 
+	load x; load y; +; store sum
+
+	returns sum as-uint32
+	`)
+
+	testProg(t, `
+		#define x a d
+		#define d c a
+		#define hey wat's up x
+		#define c woah hey
+		int 1
+		c`,
+		AssemblerMaxVersion, Expect{5, "Macro cycle discovered: c -> hey -> x -> d -> c"}, Expect{7, "unknown opcode: c"},
+	)
+
+	testProg(t, `
+		#define c +
+		#define x a c
+		#define d x
+		#define c d
+		int 1
+		c`,
+		AssemblerMaxVersion, Expect{5, "Macro cycle discovered: c -> d -> x -> c"}, Expect{7, "+ expects..."},
+	)
+
+	testProg(t, `
+		#define X X
+		int 3`,
+		AssemblerMaxVersion, Expect{2, "Macro cycle discovered: X -> X"},
+	)
+
+	// Check that macros names can't be things like named constants, opcodes, etc.
+	// If pragma is given, only macros that violate that version's stuff should be errored on
+	testProg(t, `
+		#define return random
+		#define pay randomm
+		#define NoOp randommm
+		#define + randommmm
+		#pragma version 1 // now the versioned check should activate and check all previous macros
+		#define return hi // no error b/c return is after v1
+		#define + hey // since versioned check is now online, we can error here
+		int 1`,
+		assemblerNoVersion,
+		Expect{3, "Named constants..."},
+		Expect{4, "Named constants..."},
+		Expect{6, "Macro names cannot be opcodes: +"},
+		Expect{8, "Macro names cannot be opcodes: +"},
+	)
+
+	// Same check, but this time since no version is given, the versioned check
+	// uses AssemblerDefaultVersion and activates on first instruction (int 1)
+	testProg(t, `
+		#define return random
+		#define pay randomm
+		#define NoOp randommm
+		#define + randommmm
+		int 1 // versioned check activates here
+		#define return hi
+		#define + hey`,
+		assemblerNoVersion,
+		Expect{3, "Named constants..."},
+		Expect{4, "Named constants..."},
+		Expect{6, "Macro names cannot be opcodes: +"},
+		Expect{8, "Macro names cannot be opcodes: +"},
+	)
+
+	testProg(t, `
+		#define Sender hello
+		#define ApplicationArgs hiya
+		#pragma version 1
+		#define Sender helllooooo
+		#define ApplicationArgs heyyyyy // no error b/c ApplicationArgs is after v1
```

Comment:
oof, redefining ApplicationArgs.  
Last chance to force macros to start with some set prefix?